doc/developer/design/19042023_stash_migrations.md

```diff
@@ -0,0 +1,443 @@
+        1 => migrations::v1_to_v2::migrate(...).await?;
+        2 => migrations::v2_to_v3::migrate(...).await?;
+        3 => // migrate
+        NEXT_VERSION.. => panic!("Traveled to the future!"),
+    }
+}
+```
+
+There are a few benefits to putting each migration function in their own modules:
+
+1. The only place we would import the generated protobuf code from our snapshots would be in these
+modules. So the only place that would import the `object_v1.proto` would be the `v0_to_v1` and
+`v1_to_v2` modules. This helps prevent us from depending on the snapshotted versions anywhere
+besides the necessary migration paths.
+2. We promote a pattern of documenting the migration and why it exists in a module doc comment.
+While it's possible to have doc comments on functions, I believe it would be more obvious/easier to
+review module level doc comments.
+
+Each migration function would be provided a `stash::Transaction`, which would allow us to open
+[`StashCollection`](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/transaction.rs#L136)s
+which should provide enough flexibility to facilitate any migration.
+
+> Note: `StashCollection`s do not enforce uniqueness like [`TableTransaction`](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/lib.rs#L775-L793)s
+do, some more thought is needed here as to whether or not uniqueness checks are required for
+migrations. They probably are?
+
+
+### Part 3: Initializing a Stash
+
+We add a new method to the Stash that is `async fn initialize(...) -> Result<...>`, that will
+assert the Stash is empty, and then populate it with our inital values. We can already have logic
+to detect when a Stash [has not been initialized](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/adapter/src/catalog/storage.rs#L1138-L1141).
+Instead of returning a version of `0`, we would now call this new `initialize(...)` method. The
+method itself would largely be very similar, if not the same, to our [first migration step](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/adapter/src/catalog/storage.rs#L90-L418),
+a new behavior though is it would initialize `"user_version"` to the `STASH_VERSION` defined in
+part 2.
+
+Creating this separate initialization step, allows us to deprecate old migrations, e.g. the
+`migrate(...)` function would be able to look something like:
+
+```
+const MIN_STASH_VERSION = 3;
+
+async fn migrate(...) -> Result<...> {
+    let version = stash.read_version().await?;
+
+    const NEXT_VERSION = STASH_VERSION + 1;
+    match version {
+        ..MIN_STASH_VERSION => panic!("Deprecated stash version!"),
+        3 => migrations::v3_to_v4::migrate(...).await?;
+        4 => migrations::v4_to_v5::migrate(...).await?;
+        NEXT_VERSION.. => panic!("Traveled to the future!"),
+    }
+}
+```
```

Comment:
@mjibson here is a small blurb on deprecating old migrations. The idea is that once we've released the migration, in the next release of the binary we could bump `MIN_STASH_VERSION` and remove the old migration path, similar to what we do today.

Reply:
Ah ok! LGTM then, should be ok to merge.